# QTL_MYCO
Impact of host genetics and diet on the murine intestinal mycobiome.

Yask Gupta, Anna Lara Ernst, Artem Vorobyev, Foteini Beltsiou, Detlef Zillikens, Katja Bieber, Simone Sanna-Cherchi, Angela M. Christiano, Christian D. Sadik, Ralf J. Ludwig and Tanya Sezin

SUMMARY 
The mammalian gut is a home to a diverse microbial ecosystem, whose composition affects various physiological and pathophysiological traits of the host organism. Recently, next generation sequencing-based meta-genomic approaches demonstrated how the interplay of host genetics, microbiome, and environmental factors shape complex traits and clinical outcomes. Much of the data on microbial communities within mammalian hosts has been derived from the analysis of bacterial communities, whereas other communities, such as fungi, have been understudied. More recently, sequencing of the ribosomal DNA internal transcribed spacer (ITS) of nuclear DNA has helped to reveal the biological role of fungi and their regulation by host genetics. Here, we provide in vivo evidence in mice that fungal communities are regulated by host genetics. Using whole genome sequencing and genotyping, we mapped QTL associated with various fungal species to single genes in the host. Moreover, we show that diet and itsâ€™ interaction with host genetics alter the composition of fungi in outbred mice, and we identified fungal indicator species associated with different dietary regimes. Collectively, we uncovered an association of the intestinal fungal community with host genetics, as well as a regulatory role of diet on this ecological niche.  

**INDEX**

Compositional analysis of Fungi is in script folder "CompositionAnalysisMyco.R".
Please install R libraries indicspecies, pheatmap, phyloseq, reshape2, ggplot2, vegan, ggpubr, MANOVA.RM before running the script. 
